pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -166,46 +189,206 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	var lock sync.Mutex
+
+	topologyPairsPodSpreadMap := &topologyPairsPodSpreadMap{
+		// topologyKeyToMinPodsMap will be initialized with proper size later.
+		topologyPairsMaps: newTopologyPairsMaps(),
+	}
+
+	appendTopologyPairsMaps := func(toAppend *topologyPairsMaps) {
+		lock.Lock()
+		topologyPairsPodSpreadMap.appendMaps(toAppend)
+		lock.Unlock()
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	processNode := func(i int) {
+		nodeInfo := nodeInfoMap[allNodeNames[i]]
+		node := nodeInfo.Node()
+		if node == nil {
+			klog.Errorf("node %q not found", allNodeNames[i])
+			return
+		}
+		// In accordance to design, if NodeAffinity or NodeSelector is defined,
+		// spreading is applied to nodes that pass those filters.
+		if !podMatchesNodeSelectorAndAffinityTerms(pod, node) {
+			return
+		}
+		// Ensure current node's labels contains all topologyKeys in 'constraints'.
+		for _, constraint := range constraints {
+			if _, ok := node.Labels[constraint.TopologyKey]; !ok {
+				return
+			}
+		}
+
+		nodeTopologyMaps := newTopologyPairsMaps()
+		// nodeInfo.Pods() can be empty; or all pods don't fit
+		for _, existingPod := range nodeInfo.Pods() {
+			if existingPod.Namespace != pod.Namespace {
+				continue
+			}
+			podLabelSet := labels.Set(existingPod.Labels)
+			for _, constraint := range constraints {
+				ok, err := podMatchesSpreadConstraint(podLabelSet, constraint)
+				if err != nil {
+					errCh.SendErrorWithCancel(err, cancel)
+					return
+				}
+				if ok {
+					// constraint.TopologyKey is already guaranteed to be present
+					pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
+					nodeTopologyMaps.addTopologyPair(pair, existingPod)
+				}
+			}
+		}
+		// If needed, append topology pair without entry of pods.
+		// For example, on node-x, there is no pod matching spread constraints,
+		// but node-x should be also considered as a match (with match number 0)
+		// i.e. <node: node-x>: {}
+		for _, constraint := range constraints {
+			// constraint.TopologyKey is already guaranteed to be present
+			pair := topologyPair{
+				key:   constraint.TopologyKey,
+				value: node.Labels[constraint.TopologyKey],
+			}
+			// addTopologyPairWithoutPods is a non-op if other pods match this pair
+			nodeTopologyMaps.addTopologyPairWithoutPods(pair)
+		}
+
+		appendTopologyPairsMaps(nodeTopologyMaps)
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processNode)
+
+	if err := errCh.ReceiveError(); err != nil {
+		return nil, err
+	}
+
+	// calculate min match for each topology pair
+	topologyPairsPodSpreadMap.topologyKeyToMinPodsMap = make(map[string]int32, len(constraints))
+	for _, constraint := range constraints {
+		topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[constraint.TopologyKey] = math.MaxInt32
+	}
+	for pair, podSet := range topologyPairsPodSpreadMap.topologyPairToPods {
+		// TODO(Huang-Wei): short circuit unvisited portions of <topologyKey: any value>
```

Comment:
I wonder what type of short circuiting you were thinking about.
Without short circuiting, the following comparison is performed and we continue to the next pair:
```
l < topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[pair.key]
```